File: servidor-java/src/config/HostConfiguration.java
package config;

import com.google.gson.Gson;
import com.google.gson.JsonParseException;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;


public final class HostConfiguration {
    private static final Gson GSON = new Gson();

    private String host_id;
    private Address listen;
    private Gateway gateway;
    private RemoteHost remote_host;

    public static HostConfiguration load(String configurationFilePath) {
        try {
            String configurationText = Files.readString(Path.of(configurationFilePath));
            HostConfiguration configuration = GSON.fromJson(
                configurationText,
                HostConfiguration.class
            );
            configuration.validate();
            return configuration;
        } catch (IOException | JsonParseException exception) {
            throw new IllegalArgumentException(
                "No fue posible leer la configuración del banco.",
                exception
            );
        }
    }

    public String getHostId() {
        return host_id;
    }

    public Address getListenAddress() {
        return listen;
    }

    public Gateway getGateway() {
        return gateway;
    }

    public RemoteHost getRemoteHost() {
        return remote_host;
    }

    private void validate() {
        validateText(host_id, "host_id");
        if (listen == null || gateway == null || remote_host == null) {
            throw new IllegalArgumentException("Faltan listen, gateway o remote_host.");
        }
        listen.validate("listen");
        gateway.validate();
        remote_host.validate();
    }

    private static void validateText(String value, String fieldName) {
        if (value == null || value.isBlank()) {
            throw new IllegalArgumentException(fieldName + " debe ser texto no vacío.");
        }
    }

    public static final class Address {
        private String ip;
        private int port;

        public String getIp() {
            return ip;
        }

        public int getPort() {
            return port;
        }

        private void validate(String fieldName) {
            validateText(ip, fieldName + ".ip");
            validatePort(port, fieldName + ".port");
        }
    }

    public static final class Gateway {
        private String router_id;
        private String ip;
        private int port;

        public String getRouterId() {
            return router_id;
        }

        public String getIp() {
            return ip;
        }

        public int getPort() {
            return port;
        }

        private void validate() {
            validateText(router_id, "gateway.router_id");
            validateText(ip, "gateway.ip");
            validatePort(port, "gateway.port");
        }
    }

    public static final class RemoteHost {
        private String host_id;
        private String gateway_id;

        public String getHostId() {
            return host_id;
        }

        public String getGatewayId() {
            return gateway_id;
        }

        private void validate() {
            validateText(host_id, "remote_host.host_id");
            validateText(gateway_id, "remote_host.gateway_id");
        }
    }

    private static void validatePort(int port, String fieldName) {
        if (port < 1 || port > 65535) {
            throw new IllegalArgumentException(fieldName + " debe estar entre 1 y 65535.");
        }
    }
}

File: cliente-python/src/config/host_configuration.py
import json
from dataclasses import dataclass
from pathlib import Path


@dataclass(frozen=True)
class Address:
    ip: str
    port: int


@dataclass(frozen=True)
class HostConfiguration:
    host_id: str
    listening_address: Address
    gateway_id: str
    gateway_address: Address
    remote_host_id: str
    remote_gateway_id: str


def load_host_configuration(configuration_file_path: str) -> HostConfiguration:
    try:
        configuration = json.loads(
            Path(configuration_file_path).read_text(encoding="utf-8")
        )
        return HostConfiguration(
            host_id=read_text(configuration, "host_id"),
            listening_address=read_address(configuration["listen"]),
            gateway_id=read_text(configuration["gateway"], "router_id"),
            gateway_address=read_address(configuration["gateway"]),
            remote_host_id=read_text(configuration["remote_host"], "host_id"),
            remote_gateway_id=read_text(configuration["remote_host"], "gateway_id"),
        )
    except FileNotFoundError as exception:
        raise ValueError("No existe el archivo de configuración del host.") from exception
    except (KeyError, TypeError, json.JSONDecodeError) as exception:
        raise ValueError("La configuración del host no tiene el formato esperado.") from exception


def read_address(address_data: dict) -> Address:
    port = address_data.get("port")

    if isinstance(port, bool) or not isinstance(port, int) or not 1 <= port <= 65535:
        raise ValueError("El puerto debe estar entre 1 y 65535.")

    return Address(ip=read_text(address_data, "ip"), port=port)


def read_text(json_object: dict, field_name: str) -> str:
    value = json_object.get(field_name)

    if not isinstance(value, str) or not value.strip():
        raise ValueError(f"El campo '{field_name}' debe ser texto no vacío.")

    return value
